src/render/painter.js

```diff
@@ -755,6 +755,8 @@ class Painter {
         // The default values for this state is meaningful and often expected.
         // Leaving this state dirty could cause a lot of confusion for users.
         this.context.cullFace.setDefault();
+        this.context.frontFace.setDefault();
+        this.context.cullFaceSide.setDefault();
```

Comment:
That looks good, this exhausts the missing one and I don't think it is necessary to add any other states to reset here. Just to confirm, the one not defaulted here are blending, clearColor, depthTest, depthMask, depthFunc, stencilTest, stencilMask, current program & fbo.

color modes are overridden here: https://github.com/mapbox/mapbox-gl-js/blob/ef7b2cb243f983fef7482502021e26bda93896e2/src/render/draw_custom.js#L34
stencil here: https://github.com/mapbox/mapbox-gl-js/blob/ef7b2cb243f983fef7482502021e26bda93896e2/src/render/draw_custom.js#L35
depth here: https://github.com/mapbox/mapbox-gl-js/blob/ef7b2cb243f983fef7482502021e26bda93896e2/src/render/draw_custom.js#L41

And custom layer should override the program.